controllers/internal/utils/dns.go

```diff
@@ -20,30 +20,92 @@ package utils
 import (
 	"fmt"
 	"sort"
+	"strings"
 
 	"github.com/lixiangzhong/dnsutil"
+	"github.com/miekg/dns"
 )
 
-// Dig retrieves list of tuple <IP address, A record > from edge DNS server for specific FQDN
-func Dig(edgeDNSServer, fqdn string) ([]string, error) {
+type DNSServer struct {
+	Host string
+	Port int
+}
+
+type DNSList []DNSServer
+
+func (s DNSServer) String() string {
+	return fmt.Sprintf("%s:%v", s.Host, s.Port)
+}
+
+func (l DNSList) String() string {
+	var aux []string
+	for _, el := range l {
+		aux = append(aux, el.String())
+	}
+	return strings.Join(aux, ",")
+}
+
+// Dig returns list of tuple <IP address, A record > for a given FQDN by using the dns servers from edgeDNSServers
+// dns servers are tried taking always two of them (setting one as a primary and the other as a backup one)
+// and if there is a non-error response it is returned and the rest is not tried
+func Dig(fqdn string, edgeDNSServers ...DNSServer) ([]string, error) {
 	var dig dnsutil.Dig
-	if edgeDNSServer == "" {
-		return nil, fmt.Errorf("empty edgeDNSServer")
+	if len(edgeDNSServers) == 0 {
+		return nil, fmt.Errorf("empty edgeDNSServers, provide at least one")
 	}
-	err := dig.SetDNS(edgeDNSServer)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)", edgeDNSServer, err)
-		return nil, err
+	for i := 0; i < len(edgeDNSServers); i += 2 {
+		if edgeDNSServers[i].Host == "" {
+			return nil, fmt.Errorf("empty dns server's hostname/ip")
+		}
+		err := dig.SetDNS(edgeDNSServers[i].String())
+		if err != nil {
+			err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)",
+				edgeDNSServers[i].String(), err)
+			return nil, err
+		}
+		if len(edgeDNSServers) > i+1 {
+			backupDNSServer := edgeDNSServers[i+1]
+			if backupDNSServer.Host == "" {
+				return nil, fmt.Errorf("empty edgeDNSServer")
+			}
+			err := dig.SetBackupDNS(backupDNSServer.String())
+			if err != nil {
+				err = fmt.Errorf("dig error: can't set dns backup server (%s) with error(%s)",
+					backupDNSServer.String(), err)
+				return nil, err
+			}
+		}
+		a, err := dig.A(fqdn)
+		if err != nil {
+			if len(edgeDNSServers) > i+2 { // there are still fallback dns servers listed
+				continue
+			} else {
+				err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
+				return nil, err
+			}
+		}
+		var ips []string
+		for _, ip := range a {
+			ips = append(ips, fmt.Sprint(ip.A))
+		}
+		sort.Strings(ips)
+		return ips, nil
 	}
-	a, err := dig.A(fqdn)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
-		return nil, err
+	return nil, fmt.Errorf("dig error: all dns servers were tried and none of them were able to resolve")
+}
```

Comment:
can't we just do that simpler way ? 
(find first  server responding, if ok than, provide iplist ? 
As benefit we can completelly remove dependency on `"github.com/lixiangzhong/dnsutil"`

```go
func Dig(fqdn string, edgeDNSServers ...DNSServer) (ips []string, err  error) {
	var ack *dns.Msg
	if len(edgeDNSServers) == 0 {
		return nil, fmt.Errorf("empty edgeDNS servers")
	}

	var processAck = func(ack *dns.Msg) (ips []string) {
		for _, a := range ack.Answer {
			ips = append(ips, a.(*dns.A).A.String())
		}
		sort.Strings(ips)
		return
	}

	for _, ns := range edgeDNSServers {
		r := new(dns.Msg)
		r.SetQuestion(fqdn+".", dns.TypeA)
		ack, err = dns.Exchange(r, fmt.Sprintf("%s:%v",ns.Host, ns.Port))
		if err == nil {
			return processAck(ack), nil
		}
	}

	err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
	return
}
```

Reply:
I've changed the implementation to something that looks close to yours, but using internally the `Exchange` function in the same file. It's being tried 1-by-1 from the slice -> algorithm should be easier to read

Reply:
@jkremser , thx,  I didn't check, but please clean `go.mod `from dependency on `github.com/lixiangzhong/dnsutil` if you didn't already

Reply:
gone, what about `terratest/utils/extensions.go:345` and `terratest/go.mod`, currently it depends on `github.com/AbsaOSS/gopkg v0.1.2` so it has the old/current implementation of Dig.. I assume it's kind of a chicken-egg problem, it needs to depend on a released version, but first this change needs to go in and only then we can bump the implementation also in the `terratest/utils/extensions.go`

Reply:
terratest is technically different project. After checking dependencies I see:
```
# github.com/lixiangzhong/dnsutil
k8gbterratest/utils
github.com/AbsaOSS/gopkg/dns
github.com/lixiangzhong/dnsutil
```
^^^ there is no direct dependency to `github.com/lixiangzhong/dnsutil`. 

Creating an issue in github.com/AbsaOSS/gopkg package.  We can move implementation there in the future.